### ISSUE TITLE
fix: Issue with Frequency Change in Waveform: When Changing Frequencies, Background Frequency Persists and High Frequencies Take Time to Generate

### ISSUE THREAD

**Issue by Waruna-Kaushalya** [alnitak/flutter_soloud#129]
Description

When changing the frequency of waveform sounds using the flutter_soloud package, after a while, some background frequencies are heard along with the original frequency. Additionally, when changing to high frequencies (above 10,000 Hz), the engine struggles and takes extra time to generate the sound.

Steps To Reproduce

1. Use `SoLoud.instance.loadWaveform(WaveForm.sin, true, 1, 0)` to generate a sine wave sound.
2. Change the frequency of the generated sound using: `SoLoud.instance.setWaveformFreq(soundHandle, frequency);`
3. Notice that after the frequency change, a lower frequency can be heard in the background with the original frequency.
4. High frequencies (above 10,000 Hz) take longer to generate, and there’s a delay in producing sound.

Expected Behavior

The frequency change should happen without any background or low-frequency sound. High frequencies (above 10,000 Hz) should play immediately without delay.

Please help resolve this issue.

**Comment by Sandun-Induranga** [comment]
> Description
> 
> When changing the frequency of waveform sounds using the flutter_soloud package, after a while, some background frequencies are heard along with the original frequency. Additionally, when changing to high frequencies (above 10,000 Hz), the engine struggles and takes extra time to generate the sound.
> 
> Steps To Reproduce
> 
> 1. Use `SoLoud.instance.loadWaveform(WaveForm.sin, true, 1, 0)` to generate a sine wave sound.
> 2. Change the frequency of the generated sound using: `SoLoud.instance.setWaveformFreq(soundHandle, frequency);`
> 3. Notice that after the frequency change, a lower frequency can be heard in the background with the original frequency.
> 4. High frequencies (above 10,000 Hz) take longer to generate, and there’s a delay in producing sound.
> 
> Expected Behavior
> 
> The frequency change should happen without any background or low-frequency sound. High frequencies (above 10,000 Hz) should play immediately without delay.
> 
> Please help resolve this issue.

I also face this issue

**Comment by KavinduDhananjaya** [comment]
I also facing this issue.

**Comment by Waruna-Kaushalya** [comment]
@alnitak , Thank you for the prompt reply, and I appreciate your support in resolving this issue.

### 1st Issue:

`The issue I'm experiencing is when using a frequency >10,000 Hz. The sound is not constant and has some frequency oscillations.`

Yes, you are correct. This is one of the issues we found. After 10,000 Hz, the engine struggles to generate the frequencies, and the sound becomes inconsistent with oscillations. This is one of the problems I’m facing, where the engine struggles to maintain a stable frequency.

### 2nd Issue:

This issue happens with any frequency (we tested from 100 Hz to 16,000 Hz). It occurs as follows: when we set a frequency at some point, another lower frequency is heard along with the original frequency. For example, when we set the frequency to 4500 Hz, after around 20 seconds, another lower frequency (like 3000 Hz) can be heard along with the 4500 Hz. It takes some time to reproduce, typically around 20 seconds.

### Explanation for `SoLoud.instance.play()`:

Yes, as you mentioned, if we don't stop the previous `SoundHandle` and play it again, it creates another instance of the sound, playing over the previous one. However, we are not doing that. In the second issue I mentioned, the problem occurs even when creating a single `SoundHandle` and playing it only once. This issue happens with that single instance of the audio.

### Sample videos for the issue 2:

Video one: Another frequency can be heard after 0:06

https://github.com/user-attachments/assets/1307c2f8-ae80-4ca9-81e2-752e2d9c5337

Video two: Two additional frequencies can be heard after 0:14 and 0:26.

https://github.com/user-attachments/assets/791c5ee4-1b62-4e78-9ce6-7e9ef406bb9e

### How to Reproduce Issue 2:

Here’s the sample code with a slider to change the frequency:

**Steps to Reproduce:**

1. First, tap the "Play" button.
2. Then move the slider and leave it at a point, such as 4000 Hz.
3. After about 20 seconds, you will hear another lower frequency along with the original 4000 Hz frequency.

```dart
import 'package:flutter/material.dart';
import 'package:flutter_soloud/flutter_soloud.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({required this.title, super.key});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double frequency = 1000;
  AudioSource? currentSound;
  SoundHandle? soundHandle;
  bool isPlaying = false;

  @override
  void dispose() {
    stop();
    SoLoud.instance.deinit();
    super.dispose();
  }

  Future<void> play(double frequency) async {
    try {
      if (!SoLoud.instance.isInitialized) {
        await SoLoud.instance.init();
      }

      if (isPlaying) {
        await stop();
      }

      currentSound =
          await SoLoud.instance.loadWaveform(WaveForm.sin, true, 1, 0);

      soundHandle = await SoLoud.instance.play(
        currentSound!,
        loopingStartAt: const Duration(seconds: 5),
        looping: true,
      );

      setState(() {
        isPlaying = true;
      });

      SoLoud.instance.setWaveformFreq(currentSound!, frequency);
    } catch (e) {
      debugPrint('Error while trying to play sound: $e');
    }
  }

  Future<void> stop() async {
    try {
      if (soundHandle != null && isPlaying) {
        await SoLoud.instance.stop(soundHandle!);
        setState(() {
          isPlaying = false;
        });
      }
    } catch (e) {
      debugPrint('Error while trying to stop sound: $e');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text('Freequncy : $frequency'),
            Slider(
              value: frequency,
              min: 20,
              max: 16000,
              onChanged: (value) {
                setState(() {
                  frequency = value;

                  if (currentSound != null && isPlaying) {
                    SoLoud.instance.setWaveformFreq(currentSound!, value);
                  }
                });
              },
              label: 'Frequency: ${frequency.toStringAsFixed(0)} Hz',
              activeColor: Colors.green,
              inactiveColor: Colors.green[100],
            ),
            const SizedBox(height: 50),
            ElevatedButton(
              onPressed: isPlaying ? null : () => play(frequency),
              child: const Text('Play'),
            ),
            const SizedBox(height: 20),
            ElevatedButton(
              onPressed: isPlaying ? stop : null,
              child: const Text('Stop'),
            ),
          ],
        ),
      ),
    );
  }
}

```

**Comment by alnitak** [comment]
@Waruna-Kaushalya thank you very much for the detailed description!

I think this is a hard-to-find bug inside the waveform generation from SoLoud C++ lib examples. 
Maybe I found a solution, but it will take some time. I don't know how much.

I'll leave this open till then.

**Comment by Waruna-Kaushalya** [comment]
@alnitak, thank you for your response and for looking into this. I appreciate your efforts to find a solution and will wait for your updates.

**Comment by alnitak** [comment]
"hard-to-find bugs" become easy when you find them!!

I think it's fixed now. Can you please confirm it is working on your side? Just try to use the `fix_waveform` branch in your `pubspec.yaml` by replacing:
```
flutter_soloud: ^2.1.3
```
with:
```
flutter_soloud:
  path: https://github.com/alnitak/flutter_soloud
  ref: fix_waveform
```
Probably you need to do a `flutter clean`.

**Comment by Waruna-Kaushalya** [comment]
@alnitak Sure, I'll check it and update you.

**Comment by Waruna-Kaushalya** [comment]
@alnitak  Good news! 🎉 The issue with the overlapping frequencies has been fixed. I spent some time carefully checking various frequencies, and the sound is now stable without any overlapping. Additionally, the issue with generating high frequencies has also been resolved. The audio is now generated smoothly and effortlessly. Both issues have been fully addressed, and I truly appreciate the dedication and effort you’ve put into resolving them. 🙏 Thank you so much for your hard work and support!
